### ISSUE TITLE
Inability to display some fonts

### ISSUE THREAD

**Issue by VladWinner** [notepad-plus-plus/notepad-plus-plus#9951]
### Description of the Issue
I noticed that some fonts refuse to be displayed in Notepad++ (and are replaced by system fonts), although they work fine in other programs. At the same time, most of these fonts work, but only in regular font style. So you just can't select a font with style other than regular from the drop-down list.

### Steps to Reproduce the Issue
List of fonts that don't work:
✓ Cascadia Code (Regular)
× Cascadia Code (non-regular styles)
✓ Cascadia Mono (Regular)
× Cascadia Mono (non-regular styles)
✓ Commissioner (Regular)
× Commissioner (non-regular styles)
✓ Exo 2 (Regular)
× Exo 2 (non-regular styles)
✓ Fira Code (Regular)
× Fira Code (non-regular styles)
✓ Fira Mono (Regular)
× Fira Mono (non-regular styles)
✓ Fira Sans (Regular)
× Fira Sans (non-regular styles)
✓ FiraGO (Regular)
× FiraGO (non-regular styles)
✓ Georgia Pro (Regular)
× Georgia Pro (non-regular styles)
× Golos Text (Regular, Black, DemiBold, Medium)
✓ Golos Text VF (Regular)
× Golos Text VF (Black, Demibold, Medium)
✓ IBM Plex Mono (Regular)
× IBM Plex Mono (non-regular styles)
✓ IBM Plex Sans (Regular)
× IBM Plex Sans (non-regular styles)
✓ IBM Plex Serif (Regular)
× IBM Plex Sans (non-regular styles)
✓ Inconsolota (Regular)
× Inconsolota (non-regular styles)
× Inconsolota Condensed
× Inconsolota Expanded
× Input Mono
× Input Serif
✓ Inter (Regular)
× Inter (non-regular styles)
✓ Iosevka (Regular)
× Iosevka Extended
✓ Iosevka SS** (Regular)
× Iosevka SS** Extended (any Stylistic Set)
✓ JetBrains Mono (Regular)
× JetBrains Mono (non-regular styles)
✓ Lato (Regular)
× Lato (non-regular styles)
✓ Noto Sans (Regular)
× Noto Sans (non-regular styles)
✓ Open Sans (Regular)
× Open Sans (non-regular styles)
✓ PT Root UI (Regular)
× PT Root UI (non-regular styles)
✓ Roboto (Regular)
× Roboto (non-regular styles)

At the same time, using the standard style sliders in Notepad++, you cannot select, for example, extended/expanded font style (semibold and italic only for now).

### Debug Information
Notepad++ v7.9.5   (64-bit)
Build time : Mar 21 2021 - 02:13:17
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : ON
Local Conf mode : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 2009
OS Build : 19044.1023
Current ANSI codepage : 1251
Plugins : ComparePlugin.dll DSpellCheck.dll mimeTools.dll NppConverter.dll NppExport.dll 
System Language : Russian

Administrator mode does not affect this.

**Comment by sasumner** [comment]
Does this setting have any effect?:
![image](https://user-images.githubusercontent.com/30118311/120890561-91ce8680-c5d1-11eb-9a87-f8e435eb84e7.png)

**Comment by VladWinner** [comment]
> Does this setting have any effect?:
> ![image](https://user-images.githubusercontent.com/30118311/120890561-91ce8680-c5d1-11eb-9a87-f8e435eb84e7.png)

Yes, disabling the checkbox fixes the issue.

**Comment by sasumner** [comment]
It's really all controlled by Scintilla; not directly controllable by Notepad++.

**Comment by VladWinner** [comment]
https://sourceforge.net/p/scintilla/bugs/2262/

**Comment by mnissl** [comment]
> It's really all controlled by Scintilla; not directly controllable by Notepad++.

No, it's up to the application to control Scintilla correctly. I ran into the same situation with my application which uses Scintilla when switching from GDI to DirectWrite.

I found out that GDI can handle font names with weight in their name, like "Cascadia Mono SemiLight", but DirectWrite cannot.

DirectWrite wants you to pass in the family name "Cascadia Mono" and the corresponding weight (350 = SemiLight).

As a reference implementation, see how Vim does it correctly in [DWriteContext::CreateTextFormatFromLOGFONT()](https://github.com/vim/vim/blob/master/src/gui_dwrite.cpp#L733).

**Comment by oldium** [comment]
@sasumner It looks like that this is not Scintilla-specific, but Notepad++-specific issue, so the “scintilla dependent” label on this issue is not correct.

**Comment by mnissl** [comment]
> @sasumner It looks like that this is not Scintilla-specific, but Notepad++-specific issue, so the “scintilla dependent” label on this issue is not correct.

Yes and no. GDI can handle font names that contain weight (like "Cascadia Mono SemiLight") whereas DirectWrite requires to specify the family ("Cascadia Mono") and the weight (350 = SemiLight) separately.

The Scintilla documentation does not point out this difference. I had to find out myself by digging into the source code. It ain't surprising that everybody else runs into this issue as well.

In that regard, I provided a patch which enhances Scintilla to work things out for DirectWrite: https://sourceforge.net/p/scintilla/code/merge-requests/36/

Alas, my patch was torn down by Neil: https://sourceforge.net/p/scintilla/bugs/2356/

I use my patch successfully in my app where the configuration dialog also lists all available font faces in a combobox, just like Notepad++ does.

So my patch might be an option for Notepad++ as well to make all font faces available when using DirectWrite while not having to redo the font configuration dialog.

**Comment by alankilborn** [comment]
>Alas, my patch was torn down

>my patch might be an option for Notepad++ as well to make all font faces available when using DirectWrite while not having to redo the font configuration dialog.

Probably you are reluctant to do the work of actually applying your efforts to Notepad++.
I can understand that.

**Comment by mnissl** [comment]
> Probably you are reluctant to do the work of actually applying your efforts to Notepad++. I can understand that.

[My patch]( https://sourceforge.net/p/scintilla/code/merge-requests/36/) needs to be applied to Scintilla, not to Notepad++. All you (i.e. the Notepad++ developers) need to do is ship Notepad++ with a patched version of Scintilla, not with the vanilla version. And it will just *work out of the box*. I didn't understand why Neil didn't (want to) understand.

**Comment by alankilborn** [comment]
>needs to be applied to Scintilla, not to Notepad++

Ah, OK, makes sense.

>ship Notepad++ with a patched version of Scintilla, not with the vanilla version.

Although I certainly don't make decisions here (just try to help out), I don't think this kind of "patching" is done, as it creates a maintenance problem for the future.

Although I have no idea about the validity of your patch, I'll assume it is fine and offer my sympathies that it wasn't integrated.  It's the end user that loses, and will probably seek out other alternatives for text editors that don't handle their fonts well.

**Comment by mnissl** [comment]
> Although I have no idea about the validity of your patch, I'll assume it is fine and offer my sympathies that it wasn't integrated.

My patch is fine: it also fixes a bug with the current implementation where it fails to create an HFONT handle from a DirectWrite font object. My patch was not torn down due to bad programming, but due to "ideological" reasons, as you can read [in the thread of the bug report](https://sourceforge.net/p/scintilla/bugs/2356/).

> It's the end user that loses, and will probably seek out other alternatives for text editors that don't handle their fonts well.

Exactly 😥

**Comment by oldium** [comment]
I understand the point of Scintilla maintainer. The `ChooseFont` call is not part of Scintilla, so Scintilla should not responsible to translate font representations from GDI [RBIZ font model](https://learn.microsoft.com/en-us/windows/win32/directwrite/font-selection#rbiz-font-family-model) to DirectWrite [WWS font model](https://learn.microsoft.com/en-us/windows/win32/directwrite/font-selection#weight-stretch-style-font-family-model)

**Comment by mnissl** [comment]
On the one hand, Scintilla should not be responsible. But on the other hand: let's call it a feature of Scintilla!

Especially as **everybody** gets it wrong. Especially as APIs like `ChooseFont` are used very widely: you see comboboxes with dozens of font faces (faces, not families!) in every second application. Now, every developer who uses Scintilla -- including the Notepad++ folks, including me and many more -- need to be aware of the technical details and needs to reinvent the wheel.

If Scintilla simply offered that feature, a lot of bug reports could be closed.

**Comment by oldium** [comment]
I think what we miss is actually more flexible `ChooseFont` implementation, which works with font faces, weight, width etc. This GDI version is simply too old for modern applications and extending its life does not bring better typography to applications, which is really the whole point of switching to DirectWrite.

**Comment by mnissl** [comment]
Agreed. BUT: we can lament about it or go for very sophisticated new Scintilla APIs as Neil suggested (to specify stretch and style) which require lots of time and work to implement on every platform (Windows, Linux, Mac) ... and then still *every* developer needs to adopt the new APIs ...

... or just fix the darn situation for DirectWrite with my patch.

**Comment by oldium** [comment]
Your Scintilla patch is actually a step backwards, because you are interpreting the values coming into the new DirectWrite API in a GDI fashion. Anyone who wants to take full advantage of the new API will not be able to do so. Some might argue that _face_ should be called _family_, but that's the downside of _not_ introducing the feature as cross-platform.

**Comment by mnissl** [comment]
> Your Scintilla patch is actually a step backwards, because you are interpreting the values coming into the new DirectWrite API in a GDI fashion. Anyone who wants to take full advantage of the new API will not be able to do so. Or am I reading it wrong?

The GDI interop function translates from GDI to DirectWrite.

If in future Scintilla offered APIs to specify stretch (_extended_, _condensed_, etc.) and style (_italic_, _oblique_) alongside weight (which Scintilla already offers), then just make sure that you offer your users a dialog where they can select font families (!), weight, stretch and style ... and as a developer, use the new APIs to tell Scintilla which kind of font face you would like to use eventually.

As long as Scintilla does not offer those APIs **and** as long as Notepad++ continues to allow users to select a font face (!), Notepad++ users won't be able to use DirectWrite with fonts in non default weights, stretches and styles.

**Comment by alankilborn** [comment]
@mnissl sounds like he knows what he's talking about -- a pity no one is listening... :-(

**Comment by oldium** [comment]
The DirectWrite API is not a compatibility layer for GDI. Proposal to properly extend font properties is really a way to go. This is a message from maintainer as I understand it. @mnissl is a workaround for the current situation, but it makes it hard or even impossible to change in the future at the same time.

**Comment by mnissl** [comment]
> @mnissl's patch is a workaround for the current situation (and it surely improves the current situation), but it makes it hard or even impossible to change in the future at the same time by freezing the status-quo.

This statement is wrong: Right now, you can call `SCI_STYLESETFONT` with `Cascadia Mono SemiLight` ... and it works fine with GDI, but it fails with DirectWrite (because DirectWrite says: "There is no such font family Cascadia Mono SemiLight"). My patch fixes this situation because it resolves the font family (Cascadia Mono) and weight (SemiLight) and correctly hands this information over to DirectWrite.

My patch does not prevent you from calling `SCI_STYLESETFONT` with `Cascadia Mono` and additionally calling `SCI_STYLESETWEIGHT` with `350`.

So my patch is also future compatible if Scintilla ever offers new APIs like `SCI_STYLESETSTRETCH` and `SCI_STYLESETSTYLE`.

The only thing Notepad++ needs to take care is to always call `SCI_STYLESETFONT` with font FAMILIES, never with FACES.

**Comment by oldium** [comment]
That is actually the problem, you are starting to interpret the value.

**Comment by mnissl** [comment]
> That is actually the problem, you are starting to interpret the value.

No: the actual problem is that the Scintilla documentation does not explicitly say whether `SCI_STYLESETFONT` expects a face or a family. The documentation says "font name".

At first, only GDI was available as a technolgy. And all Scintilla developers called the API above with font faces ... which by chance was understood by GDI.

Later, Scintilla supported DirectWrite, but said nothing about the fact that in this case, the API above only supports font families. And bang, there they are: all the bug reports like this one here: "Inability to display some fonts"

And now you're telling me that the problem is that my patch is interpreting the value?

All I can say is: I don't care anymore. If Scintilla is too stubborn for my patch and if the rest of Scintilla users don't understand that they cannot simply hand in font faces when using DirectWrite ... then is not my problem: Consolas in default weight, stretch and style is not that bad looking either ... and there are other great text editors in the wild which are not based on Scintilla.

**Comment by mnissl** [comment]
In order to recapulate things and answer again the question whether this DirectWrite issue is Scintilla and/or Notepad++ specific:

1. Scintilla does not offer (yet) APIs to set stretch (_extended_, _condensed_, etc.) or style (_italic_, _oblique_). Hence, fonts with such faces cannot be used.
2. Scintilla does offer an API to set weight (_bold_, _semilight_, etc.).

For point 2 to work, the font family has to be set with `SCI_STYLESETFONT` and the font weight with `SCI_STYLESETWEIGHT`.

If a font face is set with `SCI_STYLESETFONT`, DirectWrite won't be able to find the correct font family, so Scintilla falls back to its default font, Verdana.

If Notepad++ wants to enable its users to use e.g. `Cascadia Mono SemiLight` with DirectWrite, the style configuration dialog has to be redone as showing all font faces in a combobox is not compatible with GDI **and** DirectWrite at the same time.

Then again, the gain with this approach is quite little (at short term) as still a lot of fonts are left behind. In the long term, Scintilla _might_ offer those other APIs ... but the question is _if_ and _when_.

Needless to say that my patch fixes this situation right now and is compatible with future font APIs.

**Comment by alankilborn** [comment]
A perfect example of users not understanding...being disappointed that it doesn't "just work":  https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13292#issuecomment-1455022012

**Comment by GabrielBoehme13** [comment]
I'm a new user coming into Notepad++ from Notepad3, which is also Scintilla-based. Notepad3 *is* able to display some of the listed "problem" font variants with ligatures (such as JetBrains Mono SemiBold, which I use for programming).

So from that perspective, this doesn't seem to be a Scintilla issue, since it works "out of the box" for Notepad3 users once the correct font is selected ([see this issue response](https://github.com/rizonesoft/Notepad3/issues/4643#issuecomment-1464205745)).

I discovered pretty quickly that the SemiBold font variant wasn't going to work in Notepad++ if I wanted ligatures. Since Notepad3 is also open source, perhaps there's an opportunity to see how they've chosen to tackle this issue, and determine if their solution would also work for Notepad++?

**Comment by 80zhangyan** [comment]
Today, I once again encountered the problem of incorrect display of this font, I want to use alibaba font and failed.

**Comment by oldium** [comment]
Notepad3 uses local workaround, which author of Notepad++ refused to implement, because it is more like a hack than a real solution. Scintilla does not (or did not, I have not checked recently) have support to differentiate between font faces and font families (if I remember correctly - in short, DirectWrite needs more info to initialize fonts than Scinilla provides, so a custom font dialog is necessary). Anyway, the result is that users of Notepad3 have this working for years, while Notepad++ does not.
